#### <b>*OpenCV Introduction:* </b>

OpenCV is an open-source library for computer vision and image processing, ideal for Python users starting with tasks like reading images, applying filters, and object detection. It powers real-world applications from facial recognition to video analysis

*USED FOR*
- Face detection
- Object detection
- Image filtering
- Video processing
- Autonomous vehicles
- Medical imaging
- Security systems

**Computer Vision**

Computer vision is the field of AI that allows machines to *extract meaningful information from images or videos*

**Image Representation**

A computer sees images as a *matrix of numbers*

Image = Matrix (Height × Width × Channels)  ->   224 × 224 × 3

*OpenCV uses BGR format*  ->   [255,0,0] = Blue

*BGR vs RGB* — very important: OpenCV reads color images in BGR order (Blue, Green, Red), not the RGB order that most other tools use. This means the first channel of every pixel is Blue, the second is Green, the third is Red. If you display an OpenCV image using matplotlib without converting, the colors will look wrong (red objects appear blue and vice versa).

In [1]:
import cv2
import numpy as np

In [2]:
print('opencv version: ', cv2.__version__)

opencv version:  4.13.0


*image = cv2.imread(filename, flag)*

Parameter 1 — filename: A string containing the path to your image file. This can be a simple filename like 'photo.jpg' if the image is in the same folder as your Python script, or a full path like 'C:/Users/Naja/Pictures/photo.jpg'. OpenCV supports formats like JPG, PNG, BMP, TIFF, and more.

Parameter 2 — flag: Tells OpenCV how to read the image. There are three main options:

**cv2.IMREAD_COLOR** — reads the image in full color (BGR format). This is the default, so if you skip the flag, this is what happens. The value of this constant is 1.

In [3]:
img_clr= cv2.imread('apple.png',cv2.IMREAD_COLOR)
print(img_clr)

[[[239 239 239]
  [255 255 255]
  [255 255 255]
  ...
  [255 255 255]
  [255 255 255]
  [239 239 239]]

 [[255 255 255]
  [255 255 255]
  [255 255 255]
  ...
  [255 255 255]
  [255 255 255]
  [255 255 255]]

 [[255 255 255]
  [255 255 255]
  [255 255 255]
  ...
  [255 255 255]
  [255 255 255]
  [255 255 255]]

 ...

 [[255 255 255]
  [255 255 255]
  [255 255 255]
  ...
  [255 255 255]
  [255 255 255]
  [255 255 255]]

 [[239 239 239]
  [255 255 255]
  [255 255 255]
  ...
  [255 255 255]
  [255 255 255]
  [239 239 239]]

 [[ 80  80  80]
  [239 239 239]
  [255 255 255]
  ...
  [255 255 255]
  [239 239 239]
  [ 80  80  80]]]


**cv2.IMREAD_GRAYSCALE** — reads the image and immediately converts it to grayscale (single channel). The value of this constant is 0.

In [4]:
img_gray= cv2.imread('apple.png',cv2.IMREAD_GRAYSCALE)
print(img_gray)

[[239 255 255 ... 255 255 239]
 [255 255 255 ... 255 255 255]
 [255 255 255 ... 255 255 255]
 ...
 [255 255 255 ... 255 255 255]
 [239 255 255 ... 255 255 239]
 [ 80 239 255 ... 255 239  80]]


In [5]:
cv2.imshow('APPLE',img_clr)     # Display image
cv2.imshow('APPLE GRAY',img_gray)
cv2.waitKey(0)                # Wait until key pressed
cv2.destroyAllWindows()       # Close windows

*cv2.imshow()* does NOT pause your program. If you call it and nothing else, the window will flash and disappear immediately because Python moves to the next line. This is why you always call cv2.waitKey() right after.

*key = cv2.waitKey(delay)* :  delay: Time in milliseconds to wait. If you pass 0, it waits forever until any key is pressed. If you pass a number like 3000, it waits 3 seconds and then automatically continues even if no key is pressed.

*cv2.destroyAllWindows()* :
This function closes all the OpenCV windows that are currently open and frees the memory they were using.

It takes no parameters. If you want to close only one specific window, use cv2.destroyWindow('window_name') instead.

Always call this at the end of your script. Without it, windows may remain open or freeze after the program ends, especially on Windows OS.

**cv2.IMREAD_UNCHANGED** — reads the image exactly as it is, including the alpha (transparency) channel if it has one. Used for PNG files with transparent backgrounds. The value of this constant is -1.

*Return value*: cv2.imread() returns a NumPy array if successful. If the file is not found or the path is wrong, it returns None — it does NOT raise an error, it silently returns None. This is why checking if img is None is important.

In [6]:
print(img_clr.shape)

(391, 372, 3)


here the image is:
- 391 pixels tall(height)
- 372 pixels wide (width)
- 3 channels (BGR)

In [7]:
print(img_clr.dtype)

# uint8 -> pixel values range 0–255.

uint8


#### Converting to gray scale:

In [8]:
gray=cv2.cvtColor(img_clr,cv2.COLOR_BGR2GRAY)
cv2.imshow('Gray image',gray)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [9]:
cv2.imwrite('grayApple.jpg',gray)

True

In [10]:
print(img_clr[100,200])

[105 100 211]


output gives B G R values


#### Changing pixel values:

In [11]:
#image[100,200] = [255,255,255]

#### Extracting Channels

In [12]:
b,g,r = cv2.split(img_clr)
cv2.imshow("Blue", b)
cv2.imshow("Green", g)
cv2.imshow("Red", r)

cv2.waitKey(0)

-1

#### Merging Channels

In [13]:
merged= cv2.merge((b,g,r))
#cv2.imshow('merged',merged)

#### Region of interest:

In [14]:
roi= img_clr[200:400,300:200]